livekit-plugins/livekit-plugins-google/livekit/plugins/google/llm.py

```diff
@@ -0,0 +1,298 @@
+                    name=fnc_info.name,
+                    description=fnc_info.description,
+                    parameters=_build_parameters(fnc_info.arguments),
+                )
+            ]
+        )
+        for fnc_info in fnc_ctx.ai_functions.values()
+    ]
+    return tools
+
+
+def _build_parameters(arguments: Dict[str, llm.FunctionArgInfo]) -> Dict[str, Any]:
+    properties = {
+        arg_name: {
+            "type": JSON_SCHEMA_TYPE_MAP.get(arg_info.type, "string"),
+            "description": arg_info.description,
+            **({"enum": arg_info.choices} if arg_info.choices else {}),
+        }
+        for arg_name, arg_info in arguments.items()
+    }
+    required = [
+        arg_name
+        for arg_name, arg_info in arguments.items()
+        if arg_info.default is inspect.Parameter.empty
+    ]
+
+    parameters: Dict[str, Any] = {
+        "type": "object",
+        "properties": properties,
+    }
+    if required:
+        parameters["required"] = required
+
+    return parameters
+
+
+class LLMStream(llm.LLMStream):
+    def __init__(
+        self,
+        llm: LLM,
+        *,
+        vai_stream: Awaitable[AsyncIterator[GenerationResponse]],
+        chat_ctx: llm.ChatContext,
+        fnc_ctx: Optional[llm.FunctionContext],
+    ) -> None:
+        super().__init__(llm, chat_ctx=chat_ctx, fnc_ctx=fnc_ctx)
+        self._awaitable_vai_stream = vai_stream
+        self._vai_stream: Optional[AsyncIterator[GenerationResponse]] = None
+        self._function_calls_info: list[llm.FunctionCallInfo] = []  # type: ignore
+
+    async def _main_task(self) -> None:
+        if not self._vai_stream:
+            self._vai_stream = await self._awaitable_vai_stream
+        try:
+            async for response in self._vai_stream:
+                for idx, candidate in enumerate(response.candidates):
+                    chat_chunk = self._parse_candidate(candidate, idx)
+                    if chat_chunk:
+                        self._event_ch.send_nowait(chat_chunk)
+        except Exception as e:
+            logger.error("Error during main task: %s", e)
+            raise
+
+    def _parse_candidate(
+        self, candidate: Candidate, index: int
+    ) -> Optional[llm.ChatChunk]:
+        choice_delta = llm.ChoiceDelta(role="assistant")
+
+        if candidate.function_calls and self._fnc_ctx is not None:
+            for function_call in candidate.function_calls:
+                function_call_info = create_ai_function_info(
+                    fnc_ctx=self._fnc_ctx,
+                    tool_call_id=f"{function_call.name}_{uuid()}",
+                    fnc_name=function_call.name,
+                    raw_arguments=json.dumps(function_call.args),
+                )
+                self._function_calls_info.append(function_call_info)
+            choice_delta.tool_calls = self._function_calls_info
+            return llm.ChatChunk(
+                request_id=str(uuid()),
```

Comment:
Can you use `utils.shortuuid()`
Also it seems like we're generating a new request_id for each chunk?
It should be the same for the entire request